src/cmap/wait_queue.rs

```diff
@@ -1,177 +1,85 @@
 use std::{
-    collections::VecDeque,
-    sync::{Arc, Condvar, Mutex, MutexGuard},
     time::Duration,
+    convert::TryInto,
 };
 
-use derivative::Derivative;
+use futures_intrusive::sync::{SemaphoreReleaser, Semaphore};
 
 use crate::{
     error::{ErrorKind, Result},
     options::StreamAddress,
+    RUNTIME,
 };
 
 /// The wait queue ensures that threads acquiring connections proceed in a first-come, first-serve
-/// order. We wrap the internal state in an both an `Arc` and a `Mutex`; the `Arc` allows us to
-/// share the state across threads, and the mutex allows us to predicate access to the wait queue on
-/// a `Condvar`.
-///
-/// RUST-208 Investigate options in the ecosystem for a fair semaphore to use in place of the wait
-/// queue.
+/// order.
 #[derive(Debug)]
 pub(crate) struct WaitQueue {
-    inner: Arc<Mutex<WaitQueueInner>>,
-}
-
-/// The internal state of the wait queue.
-#[derive(Derivative)]
-#[derivative(Debug)]
-struct WaitQueueInner {
-    /// The elements in the queue are conditional variables. When a thread enters the wait queue,
-    /// they block on a newly-created conditional variable until either they are at the front of
-    /// the queue or an optional timeout is reached.
-    queue: VecDeque<Arc<Condvar>>,
+    max_handles: usize,
 
-    /// The timeout signifying how long a thread should wait in the queue before returning an
-    /// error. This will be the `wait_queue_timeout` for a given connection pool.
-    timeout: Option<Duration>,
+    /// A fair counting semaphore whose count corresponds to the number of connections available.
+    semaphore: Semaphore,
 
-    /// The address that the connection pool's connections will connect to. This is needed to
-    /// return a WaitQueueTimeoutError when the timeout has elapsed.
+    /// The address of the server whose pool this `WaitQueue` belongs to.
```

Comment:
Is there a reason that the additional information in the comments for `address` and `timeout` was removed?

Reply:
Nope, I think I got rid of them and then tried to rewrite them from memory or something. Reverted them back. I also added a docstring for `max_handles`.